### PR TITLE
Expose runtime storage

### LINKED ISSUE #1677: Expose runtime storage internally
### Issue To Be Solved

Embedders of Cadence would like to get access to the runtime storage, e.g. to get access to the storage map of an account and inspect stored data, modify it, etc.

### Suggested Solution

- We can easily add a function `Storage(context Context) *runtime.Storage` to `runtime.Runtime`, which e.g. gives access to storage maps via `runtime.Storage.GetStorageMap`
- `StorageMap`'s API is fairly "low-level", e.g. `ReadValue` just accepts a `key string`, not a high-level path
- `interpreter.StorageMap.ReadValue` could be changed to just accept a memory gauge, at that point, an interpreter is not required
- However, most read operations on the returned `interpreter.Value`, e.g. reading a field from a composite value through `intererpreter.CompositeValue.GetField` require an interpreter instance. Such read cases could also be changed to just accept a memory gauge
- However, mutating operations, e.g. writing a field of a composite value through `interpreter.CompositeValue.SetMember` require an interpreter with configured storage
- Should we also return a "dummy" interpreter from `Runtime.Storage(Context)` which can be used for this purpose? It should not be used for anything else though, which would make such an API brittle

### PR DESCRIPTION
**Pull request by turbolent**
Closes #1677

See linked issue

______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: runtime/runtime.go [hunk 1]
```diff
@@ -106,6 +106,9 @@ type Runtime interface {
 	// ReadLinked dereferences the path and returns the value stored at the target
 	//
 	ReadLinked(address common.Address, path cadence.Path, context Context) (cadence.Value, error)
+
+	// Storage returns the storage system.
+	Storage(context Context) *Storage
```
Comment (sideninja) [inline runtime/runtime.go:111]: excellent, I will use this branch to test out the implementation.

**Comment by sideninja** [comment]
I made implementation of this on the PR https://github.com/onflow/flow-emulator/pull/156 
I think this works great. The only thing that it probably isn't related to this PR is the conversion from interpreter values to the runtime cadence values, I noticed there is method in the runtime `ExportValue` but that requires the interpreter which I don't have access to. Should we convert the values in the first place? I'm thinking yes since runtime values are the ones I see as cadence external values that get exchanged. Any idea there?

**Comment by turbolent** [comment]
[cb86849](https://github.com/onflow/cadence/pull/1679/commits/cb868499029f5c695be2993e929ef835774635e0) also returns an interpreter which has storage configured. This allows read/write functions of `interpreter.Value` to be called which need an interpreter instance

**Comment by sideninja** [comment]
> [cb86849](https://github.com/onflow/cadence/pull/1679/commits/cb868499029f5c695be2993e929ef835774635e0) also returns an interpreter which has storage configured. This allows read/write functions of `interpreter.Value` to be called which need an interpreter instance

Great, thank you. Just a question, since exposing the interpreter requires the warning do you think we could maybe return a getter function that has the interpreter as part of the closure and just gives an option to convert a value?

**Comment by sideninja** [comment]
Something like:

```go
func (r *interpreterRuntime) Storage(context Context) (*Storage, func(value interpreter.Value) (cadence.Value, error), error) {
        // ....

	Export := func(value interpreter.Value) (cadence.Value, error) {
		return ExportValue(value, inter)	
	}
	
	return storage, Export, nil
}
```

**Comment by sideninja** [comment]
I see now you use the return interpreter in other functions on Storage. My point might be moot then

**Comment by sideninja** [comment]
I'm getting a panic, for some reason the import location handler is nil for location here https://github.com/onflow/cadence/blob/b5fbf4484e1310f2c5c8929f326a6440ed045d37/runtime/interpreter/interpreter.go#L2600 
The location in my case is `f8d6e0586b0a20c7.LockedTokens` and it crashes with:

```
2022/06/06 13:17:33 http: panic serving [::1]:55155: runtime error: invalid memory address or nil pointer dereference
goroutine 24 [running]:
net/http.(*conn).serve.func1()
	/usr/local/go/src/net/http/server.go:1825 +0xb0
panic({0x101563060, 0x102034500})
	/usr/local/go/src/runtime/panic.go:844 +0x258
github.com/onflow/cadence/runtime/interpreter.(*Interpreter).EnsureLoaded.func1()
	/Users/dapper/Dev/cadence/runtime/interpreter/interpreter.go:2604 +0x2c
github.com/onflow/cadence/runtime/interpreter.(*Interpreter).ensureLoadedWithLocationHandler(0x14001390180, {0x101731ce8, 0x1400138c078}, 0x140013c4c08)
	/Users/dapper/Dev/cadence/runtime/interpreter/interpreter.go:2627 +0xa0
github.com/onflow/cadence/runtime/interpreter.(*Interpreter).EnsureLoaded(0x14001390180?, {0x101731ce8?, 0x1400138c078?})
	/Users/dapper/Dev/cadence/runtime/interpreter/interpreter.go:2601 +0x118
github.com/onflow/cadence/runtime/interpreter.(*Interpreter).getElaboration(0x14001390180, {0x101731ce8, 0x1400138c078})
	/Users/dapper/Dev/cadence/runtime/interpreter/interpreter.go:4194 +0x2c
github.com/onflow/cadence/runtime/interpreter.(*Interpreter).getUserCompositeType(0x73014e2fc0?, {0x101731ce8?, 0x1400138c078?}, {0x14000ac6240, 0x34})
	/Users/dapper/Dev/cadence/runtime/interpreter/interpreter.go:4241 +0x30
github.com/onflow/cadence/runtime/interpreter.(*Interpreter).GetCompositeType(0x140013c4d88?, {0x101731ce8?, 0x1400138c078?}, {0x1400082e1e0?, 0x140013940d0?}, {0x14000ac6240?, 0x140008b2930?})
	/Users/dapper/Dev/cadence/runtime/interpreter/interpreter.go:4237 +0xb0
github.com/onflow/cadence/runtime/interpreter.(*Interpreter).ConvertStaticToSemaType.func2({0x101731ce8?, 0x1400138c078?}, {0x1400082e1e0?, 0x1400138c078?}, {0x14000ac6240?, 0x140013c4f98?})
	/Users/dapper/Dev/cadence/runtime/interpreter/interpreter.go:4176 +0x50
github.com/onflow/cadence/runtime/interpreter.ConvertStaticToSemaType({0x0, 0x0}, {0x101734528?, 0x140008b2930?}, 0x14000180700?, 0x1400138c030?)
	/Users/dapper/Dev/cadence/runtime/interpreter/statictype.go:714 +0x398
github.com/onflow/cadence/runtime/interpreter.ConvertStaticToSemaType({0x0, 0x0}, {0x101734720?, 0x140008b2960?}, 0x7?, 0x1a?)
	/Users/dapper/Dev/cadence/runtime/interpreter/statictype.go:765 +0x404
github.com/onflow/cadence/runtime/interpreter.(*Interpreter).ConvertStaticToSemaType(0x1017217e8?, {0x101734720?, 0x140008b2960?})
	/Users/dapper/Dev/cadence/runtime/interpreter/interpreter.go:4169 +0x70
github.com/onflow/cadence/runtime/interpreter.(*Interpreter).MustConvertStaticToSemaType(...)
	/Users/dapper/Dev/cadence/runtime/interpreter/interpreter.go:4182
github.com/onflow/cadence/runtime.exportLinkValue({{0x1, {0x140013a8120, 0x1a}}, {0x101734720, 0x140008b2960}}, 0x100547360?)
	/Users/dapper/Dev/cadence/runtime/convertValues.go:582 +0x84
github.com/onflow/cadence/runtime.exportValueWithInterpreter({0x1017394c0?, 0x140008b2ae0?}, 0x14001390180, 0x140013c5588)
	/Users/dapper/Dev/cadence/runtime/convertValues.go:176 +0xca4
github.com/onflow/cadence/runtime.ExportValue(...)
	/Users/dapper/Dev/cadence/runtime/convertValues.go:40
github.com/onflow/flow-emulator.NewAccountStorage.func1(0x0?)
	/Users/dapper/Dev/flow-emulator/accountStorage.go:57 +0x1a8
github.com/onflow/flow-emulator.NewAccountStorage({0xf8, 0xd6, 0xe0, 0x58, 0x6b, 0xa, 0x20, 0xc7}, 0x10173f840?, 0x140013880b0?)
	/Users/dapper/Dev/flow-emulator/accountStorage.go:70 +0x70
github.com/onflow/flow-emulator.(*Blockchain).GetAccountStorage(0x14000551180, {0xf8, 0xd6, 0xe0, 0x58, 0x6b, 0xa, 0x20, 0xc7})
	/Users/dapper/Dev/flow-emulator/blockchain.go:1002 +0xe8
github.com/onflow/flow-emulator/server/backend.(*Backend).GetAccountStorage(0x14001054c20, {0xf8, 0xd6, 0xe0, 0x58, 0x6b, 0xa, 0x20, 0xc7})
	/Users/dapper/Dev/flow-emulator/server/backend/backend.go:589 +0xa0
github.com/onflow/flow-emulator/server.EmulatorApiServer.Storage({0x1400019a6c0?, 0x14001076a80?, 0x14001054c20?, 0x14000559b70?}, {0x10172dd10, 0x140013ae000}, 0x1016ecb40?)
	/Users/dapper/Dev/flow-emulator/server/emulatorApi.go:143 +0x128
net/http.HandlerFunc.ServeHTTP(0x14000578a00?, {0x10172dd10?, 0x140013ae000?}, 0x1400078c051?)
	/usr/local/go/src/net/http/server.go:2084 +0x3c
github.com/gorilla/mux.(*Router).ServeHTTP(0x1400019a6c0, {0x10172dd10, 0x140013ae000}, 0x14000578900)
	/Users/dapper/go/pkg/mod/github.com/gorilla/mux@v1.8.0/mux.go:210 +0x1a0
github.com/onflow/flow-emulator/server.EmulatorApiServer.ServeHTTP(...)
	/Users/dapper/Dev/flow-emulator/server/emulatorApi.go:61
net/http.(*ServeMux).ServeHTTP(0x0?, {0x10172dd10, 0x140013ae000}, 0x14000578900)
	/usr/local/go/src/net/http/server.go:2462 +0x144
net/http.serverHandler.ServeHTTP({0x140008b2300?}, {0x10172dd10, 0x140013ae000}, 0x14000578900)
	/usr/local/go/src/net/http/server.go:2916 +0x3fc
net/http.(*conn).serve(0x14001398000, {0x10172e8a0, 0x140008b2120})
	/usr/local/go/src/net/http/server.go:1966 +0x56c
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:3071 +0x450
exporting value Link<&A.f8d6e0586b0a20c7.LockedTokens.TokenAdminCollection>(/storage/lockedTokenAdminCollection)
```

**Comment by turbolent** [comment]
@sideninja I was hoping that we could just an interpreter which only has storage configured, but yeah, looks like other options, like how to resolve imports (e.g. implicit ones, when loading values from storage), need to be provided too.

Should be fixed in [d254cbc](https://github.com/onflow/cadence/pull/1679/commits/d254cbc4223c47562b3a88710a24a67d34cbb1e1)

**Comment by sideninja** [comment]
^^ my implementation can be found here https://github.com/onflow/flow-emulator/blob/feature/storage-inspection/blockchain.go#L986

cc @turbolent

**Comment by turbolent** [comment]
@sideninja sorry, I forgot a line (initializing the context properly), should be fixed in [14882c4](https://github.com/onflow/cadence/pull/1679/commits/14882c425a63e19e6cd0f17596cd279739a84b23)

**Review by SupunS — COMMENTED** [review]
Overall looks good to me!

One question I had was: Would it make sense to split the `Storage()` API to two, one to return just the storage, and another one to return a fully-configured interpreter?
Right now, it seems that API is used for two purposes. e.g: `(r *interpreterRuntime) ReadStored()` uses it solely to create a new interpreter.

**Review by sideninja — APPROVED** [review]
With the changes made in this PR I can achieve the storage inspection needed in the emulator. It would be good if this could be merged as soon as practicle so we can complete the inspection feature.

**Review by SupunS — APPROVED** [review]
Changes look good.

Still have this question: https://github.com/onflow/cadence/pull/1679#pullrequestreview-1017259437. I'm OK with merging if this is urgent.

**Comment by sideninja** [comment]
> Changes look good.
> 
> Still have this question: [#1679 (review)](https://github.com/onflow/cadence/pull/1679#pullrequestreview-1017259437). I'm OK with merging if this is urgent.

It is mildly urgent, meaning I can wait for you concern to get resolved (I see where you are coming from) but if there's some progress, It's been now 25 days since the comment and no progress and if that stays that way then I rather merge it.

**Comment by bluesign** [comment]
I am already using this, merging would be great.

**Pull request merged**